akka-http/src/main/java/akka/http/javadsl/server/HttpApp.java

```diff
@@ -0,0 +1,150 @@
+   * Note that this method is blocking
+   */
+  public void startServer(String host, int port, ServerSettings settings, ActorSystem system) throws ExecutionException, InterruptedException {
+    startServer(host, port, settings, Optional.of(system));
+  }
+
+  /**
+   * Start a server on the specified host and port, using provided settings and [[ActorSystem]] if present
+   * Note that this method is blocking
+   */
+  public void startServer(String host, int port, ServerSettings settings, Optional<ActorSystem> system) throws ExecutionException, InterruptedException {
+
+    final ActorSystem theSystem = system.orElseGet(() -> ActorSystem.create(Logging.simpleName(this).replaceAll("\\$", "")));
+    final ActorMaterializer materializer = ActorMaterializer.create(theSystem);
+
+    CompletionStage<ServerBinding> bindingFuture = Http
+      .get(theSystem)
+      .bindAndHandle(route().flow(theSystem, materializer),
+        ConnectHttp.toHost(host, port),
+        settings,
+        theSystem.log(),
+        materializer);
+
+    bindingFuture.handle((binding, exception) -> {
+      if (exception != null) {
+        postHttpBindingFailure(exception);
+      } else {
+        //setting the server binding for possible future uses in the client
+        serverBinding.set(binding);
+        postHttpBinding(binding);
+      }
+      return null;
+    });
+
+    try {
+      bindingFuture
+        .thenCompose(ignore -> waitForShutdownSignal(theSystem)) // chaining both futures to fail fast
+        .toCompletableFuture()
+        .exceptionally(ignored -> BoxedUnit.UNIT) // If the future fails, we want to complete normally
+        .get(); // It's waiting forever because maybe there is never a shutdown signal
+    } finally {
+      bindingFuture.thenCompose(ServerBinding::unbind).handle( (success, exception) -> {
+        postServerShutdown(exception, theSystem);
+        if (!system.isPresent()) {
+          // we created the system. we should cleanup!
+          theSystem.terminate();
+        }
+        return null;
+      });
+    }
+
+  }
+
+  /**
+   * It tries to retrieve the {@link ServerBinding} if the server has been successfully started. It throws an {@link IllegalStateException} otherwise.
+   * You can use this method to attemp to retrieve the {@link ServerBinding} at any point in time to, for example, stop the server due to unexpected circumstances.
+   */
+  ServerBinding binding() {
+    if (serverBinding.get() == null)
+      throw new IllegalStateException("Binding not yet stored. Have you called startServer?");
+    return serverBinding.get();
+  }
+
+  /**
+   * Hook that will be called just after the server termination. Override this method if you want to perform some cleanup actions after the server is stopped
+   * When problems shutting down {@code possibleFailure} contains the failure occurred. It is null otherwise
+   */
+  protected void postServerShutdown(Throwable possibleFailure, ActorSystem system) {
```

Comment:
I personally would not really care about that java guideline... We want APIs to be closely mirroring one another